docs/topics/equality.md

```diff
@@ -14,17 +14,24 @@ By convention, an expression like `a == b` is translated to:
 a?.equals(b) ?: (b === null)
 ```
 
-If `a` is not `null`, it calls the `equals(Any?)` function, otherwise (`a` is `null`) it checks that `b`
+If `a` is not `null`, it calls the `equals(Any?)` function. Otherwise (`a` is `null`), it checks that `b`
 is referentially equal to `null`.
 
 Note that there's no point in optimizing your code when comparing to `null` explicitly:
 `a == null` will be automatically translated to `a === null`.
 
 To provide a custom equals check implementation, override the
 [`equals(other: Any?): Boolean`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-any/equals.html) function.
-Functions with the same name and other signatures, like `equals(other: Foo)`, don't affect equality checks with
+Functions with the same name and other signatures (like `equals(other: Foo)`) don't affect equality checks with
 the operators `==` and `!=`.
 
+While this equality operator (`==`) is called structural equality to distinguish it from the [referential equality](#referential-equality) (`===`),
+its underlying behavior could be referential in nature.
+
+One common scenario where structural equality can exhibit referential behavior is when handling non-data classes, 
+especially those without a custom implementation of the `equals()` function. In fact, this behavior is inherent by default
+in all classes except for [data classes](data-classes.md).
```

Comment:
This is not completely clearly laid out.
By default, the structural `equals()` function is implemented using referential equality in the root `Any` class.
Then classes can override it to define actually structural equality.
Value classes always override equals implementing structural equality.
Data classes also override equals the same way, unless it is final in the parent class.

Reply:
It would also be nice to explain what "structural" actually means, i.e. if one wants to implement structural equals(), how would they do that.